### PR TITLE
Text alignment reporting and enhancements for MS Excel cells and MS Word text

### LINKED ISSUE #15206: Various issues related to text alignment in Excel
### Steps to reproduce:
* Open this Excel document: [ExcelAlignment.xlsx](https://github.com/nvaccess/nvda/files/12188578/ExcelAlignment.xlsx)
* In document formatting settings (NVDA+control+D), check the "Alignment" checkbox and validate
* Move the selected cell from B1 to B7

### Actual behavior:

The speech viewer contains the following text:
align default
Standard
align left
Left
align right
Right
align default
Centered across columns  overflowing
align justify
Justify
align default
Rep
align default
Distributed
align default
cell

### Expected behavior:

All alignments should be correctly reported, more specifically centered across columns, repeated, distributed.

Also the "default" alignment in Excel should not be reported the same way as the "default" reported in Chrome on a button, where no alignment information is available.

Thus the speech viewer should contain the following text:

align general
Standard
align left
Left
align right
Right
align centered across columns
Centered across columns  overflowing
align justify
Justify
align fill
Rep
align distributed
Distributed
align default
cell

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
2023.2beta1
#### Windows version:
Windows 10 2004 (AMD64) build 19041.388
#### Name and version of other software in use when reproducing the issue:
Excel 2016, '16.0.16529.20182'
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes, same with 2023.1 and older releases.
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### LINKED ISSUE #15220: MS Word - Justified text reported as "default" text alignment
### Steps to reproduce:
* Open a document in MS Word
* In Word options, format the current paragraph as "justified"
* Press NVDA+F twice

Note: being on Windows 10, NVDA uses legacy object model to access Word documents by default, not UIA.
### Actual behavior:
"align default" is reported in formatting information.
### Expected behavior:
"align justify" should be reported in formatting information.

Indeed, "default" does not give any useful information; moreover, at least on my system, the default alignment for paragraphs in Word is "left", not "justify".

### NVDA logs, crash dumps and other attachments:
N/A
#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
2023.2beta1
#### Windows version:
Windows 10 2004 (AMD64) build 19041.388
#### Name and version of other software in use when reproducing the issue:
Office 2016 (16.0.16529.20182)
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes, same issue with 2023.1 and 2022.x
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not tested

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
Closes #15206
Closes #15220

### Summary of the issue:

Various issues for alignment reporting:
* In MS Excel horizontal or vertical text alignment is not always correctly reported.
* In MS Word (legacy), text alignment is not correctly reported for "justify" and for "distrubute".

### Description of user facing changes
Reporting text alignment in Word and Excel is now handled better in various situations. "Default" is not reported anymore since it does not give any clear indication of the position of the text.

* In MS Excel (no UIA) the alignment is reported as per Excel's options:
  * for horizontal text alignment in cell:  "General", "Centered across columns", "Repeat" and "Distribute" is now reported instead of "Default" or nothing
  * for vertical alignment: "bottom" is reported instead of "default"
* In MS Word (no UIA):
  * "justify" is now reported instead of "Default"
  * "distribute" is now reported instead of nothing

### Description of development approach

Standardized alignments returned by `getFormatField*` functions around a string enum as done in #11897. Thus, all the interfaces are impacted.

### Testing strategy:
Test plan based on #11897, adapted to this PR.

For the following APIs/files, the following test are made:
* Check that NVDA+F is able to report text alignment
* Activate the option to report text alignment and navigate with the caret to test the transitions between the various possible text alignments.

API/files tested:
- [x] NVDAObjects/IAccessible/__init__.py:
      * Test with Chrome: `data:text/html,<p align="left">left</p><p align="right">right</p><p align="center">center</p><p align="justify">justify</p>`
        OK for left, right, center, justify and match-parent; NOK for experimental justify-all, start, end without or with this PR
      * Edge (legacy)
        OK for left, right, center, justify; NOK for experimental justify-all, start, end without or with this PR
- [x] NVDAObjects/UIA/__init__.py: test with Word (UIA); left, right, center, justify are supported; but distribute is reported as justify
- [x] NVDAObjects/window/edit.py, class ITextDocumentTextInfo: Test with Wordpad (left, right, center, justify)
- [x] NVDAObjects/window/edit.py, class EditTextInfo: Paragraph alignment is not supported by NVDA (probably not natively either; no test done.
- [x]  NVDAObjects/window/excel.py: test with Excel, cell navigation (alignment of full cell).
  * The following horizontal text alignments are reported (these are the ones available in cell formatting dialog): general, left, right, center, centered across multiple columns, justify, fill and distributed.
  * The following vertical text alignments are reported (these are the ones available in cell formatting dialog): top, centered, bottom, justitfy and distributed.
  * Note: When editing the cell's content, no formatting can be reported (neither before this PR nor with this PR).
- [x] NVDAObjects/window/winword.py: test with Word 2016 (UIA not checked): left, right, center, justify and distribute are reported
- [x] virtualBuffers/MSHTML.py: Test `ui.browseableMessage`:
      * Test case:
      `ui.browseableMessage('''<p align="left">left</p><p align="right">right</p><p align="center">center</p><p align="justify">justify</p><p align="justify-all">justify-all</p><p align="start">start</p><p align="end">end</p><div align="center"><p align="match-parent">match-parent center</p></div>''', "Test", isHtml=True)`
      * Result: no change with respect to NVDA 2023.2beta1, i.e. OK for left, right, center, justify; NOK for start, end and justify-all which are all reported as left.

Tests done with:
* MS Office 2016 (16.0.16529.20182).
* Chrome 115.0.5790.110
* Edge 114.0.1823.82

#### Special case of Firefox

A part of the work of this PR was done some time ago and was working with Firefox. Today the situation is the following:
* Reporting text alignment in Firefox 116.0 (installed) is not supported, neither in NVDA 2023.2beta1 nor with this PR
* With Firefox ESR portable 102.13.0 (portable) and NVDA 2023.2beta1, text alignment is reported for justified text, but "default" is reported for left, right and centered text.
* With Firefox ESR portable 102.13.0 (portable) and this PR, text alignment is correctly reported for left, right, center and justify.

I imagine that there may be an issue with latest Firefox versions regarding text alignment reporting. @jcsteh any information or confirmation of this?

### Known issues with pull request:

1. With Word no UIA:
Not all the available alignments have been implemented. The following are missing: wdAlignParagraphJustifyHi, wdAlignParagraphJustifyLow, wdAlignParagraphJustifyMed, wdAlignParagraphThaiJustify. They were not supported before by NVDA and do not know how to use them; maybe with some specific locale?
(see https://docs.microsoft.com/en-us/office/vba/api/word.wdparagraphalignment)
If there is a need to implement them in the future, it should be quite easy.

2. Word with UIA: "distribute" is reported as "justify", since "distribute" does not seem to be a supported alignment in UIA.

3. In Excel with UIA, text alignment is not supported at all today (in NVDA 2023.2beta1). This PR does not add the support for text alignment in Excel with UIA.

4. Text alignment reporting is not supported by NVDA in:
   * PowerPoint (see #4021)
   * Libre Office Writer (see #2204)
   * Edge with UIA
 
 This PR does not improve it.

5. For web text alignment, according to [this page](https://w3c.github.io/i18n-tests/results/text-align), `text-align="justify-all"` is not supported by any browser. NVDA reports it as "left" before and after this PR.

### Change log entries:
Bug fixes
`In Word and Excel, text alignment will be correctly reported in more situations. (#15206, #15220)`

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by zstanecic** [comment]
Hi Cyrille,

The last alignment is locale specific, see description on the provided page:

Justified according to Thai formatting layout.

So, to conclude: it is locale specific.

From: Cyrille Bougot ***@***.***> 
Sent: Friday, July 28, 2023 12:22 AM
To: nvaccess/nvda ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [nvaccess/nvda] Text alignment reporting and enhancements for MS Excel cells and MS Word text (PR #15205)

 


Link to issue number:


TBD


Summary of the issue:


Various issues for alignment reporting:

*	In MS Excel horizontal or vertical text alignment is not always correctly reported.
*	In MS Word, text alignment is not correctly reported for "justify" and for "distrubute".


Description of user facing changes


*	In MS Excel (no UIA) 

*	for horizontal text alignment: "General", "Centered across columns", "Repeat" and "Distribute" is now reported instead of "Default" or nothing
*	for vertical alignment: "bottom" is reported instead of "default"

*	MS Word (no UIA): 

*	"justify" is now reported instead of "Default"
*	"distribute" is now reported instead of nothing


Description of development approach


Standardized alignments returned by getFormatField* functions around a string enum as done in #11897 <https://github.com/nvaccess/nvda/pull/11897> .


Testing strategy:


TBD


Known issues with pull request:


1.	With Word no UIA:
Not all the available alignments have been implemented. The following are missing: wdAlignParagraphJustifyHi, wdAlignParagraphJustifyLow, wdAlignParagraphJustifyMed, wdAlignParagraphThaiJustify. They were not supported before by NVDA and do not know how to use them; maybe with some specific locale?
(see https://docs.microsoft.com/en-us/office/vba/api/word.wdparagraphalignment)
If there is a need to implement them in the future, it should be quite easy.
2.	In Excel with UIA, text alignment is not supported at all today (in NVDA 2023.2beta1). This PR does not add the support for text alignment in Excel with UIA.


Change log entries:


Bug fixes
Text alignment will be reported in more situations in Word and Excel.


Code Review Checklist:


*	[ ] Pull Request description: 

*	description is up to date
*	change log entries

*	[ ] Testing: 

*	Unit tests
*	System (end to end) tests
*	Manual testing

*	[X] API is compatible with existing add-ons.
*	[X] Documentation: 

*	User Documentation
*	Developer / Technical Documentation
*	Context sensitive help for GUI changes

*	[X] UX of all users considered: 

*	Speech
*	Braille
*	Low Vision
*	Different web browsers
*	Localization in other languages / culture than English

*	[X] Security precautions taken.

  _____  


You can view, comment on, or merge this pull request online at:


  https://github.com/nvaccess/nvda/pull/15205


Commit Summary


*	c96dd4a <https://github.com/nvaccess/nvda/pull/15205/commits/c96dd4abe8f904b7238198105dfc309cf6f39d3b>  Text alignment reporting and enhancements for MS Excel cells and MS Word text.


File Changes 


(11 <https://github.com/nvaccess/nvda/pull/15205/files>  files)

*	M nvdaHelper/remote/WinWord/Constants.h <https://github.com/nvaccess/nvda/pull/15205/files#diff-a3fa5aace87571220ebe8c28b79b52ff64e103c49e400138029b2f5303993084>  (4) 
*	M nvdaHelper/remote/winword.cpp <https://github.com/nvaccess/nvda/pull/15205/files#diff-b4054d7f47f9d99612d234a1249c5833473af254dfd2f79bee543c4334957a51>  (7) 
*	M source/NVDAObjects/IAccessible/__init__.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-d433c2172b4714edb8d185ab8701f213e2bb6718c570550527e40a9f8e100f71>  (18) 
*	M source/NVDAObjects/UIA/__init__.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-55dc913388f06ed217aee6ef58c3d69a5d9eeddd6102b50d9349897b7e9851b1>  (29) 
*	M source/NVDAObjects/window/edit.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-75bd009f6ff4404204bbbf144c81b729e6c24dcfbb16572b9b2df07b4f1c08ea>  (15) 
*	M source/NVDAObjects/window/excel.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-75488b53ceb0f076739f36d1bdc8d3c68130645365f21bb21e385d7eb553f7d8>  (58) 
*	M source/NVDAObjects/window/winword.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-cc26a739b33d229b2066ab4cb0bb0f0034603036f028f3fed689090e34ec9e79>  (4) 
*	M source/controlTypes/__init__.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-7ebb6128a37f9a77ec82d334edc7880c5065d1e39ff1e392f63069cc07dabb1e>  (6) 
*	M source/controlTypes/formatFields.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-0a899a0672bb59fd3412960091f4059cc7afe281b4de3e3d3c5e9a351da8129c>  (82) 
*	M source/speech/speech.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-a29456c11e438847c67bfe5c503ef7ca66bd0ffad420b23f7cf064505622cd45>  (53) 
*	M source/virtualBuffers/MSHTML.py <https://github.com/nvaccess/nvda/pull/15205/files#diff-3ef8f37ee0d6741798e4b3ab4076883157f334e41c1757282a4516c9c7df586a>  (18) 


Patch Links:


*	https://github.com/nvaccess/nvda/pull/15205.patch
*	https://github.com/nvaccess/nvda/pull/15205.diff

**Comment by jcsteh** [comment]
Alignment is not currently implemented in the Firefox accessibility cache. This is something we'll need to address.

File: source/virtualBuffers/MSHTML.py [hunk 1]
```diff
@@ -39,6 +39,15 @@ def _getTextPositionAttribute(self, attrs: dict) -> TextPosition:
 			log.debug(f'textPositionValue={textPositionValue}')
 			return TextPosition.BASELINE
 
+	def _getTextAlignAttribute(self, attrs: dict) -> TextAlign:
```
Comment (seanbudd) [inline source/virtualBuffers/MSHTML.py:42]:
```suggestion
	def _getTextAlignAttribute(self, attrs: Dict[str, Any]) -> Optional[TextAlign]:
```

or even?
```suggestion
	def _getTextAlignAttribute(self, attrs: Dict[str, str]) -> Optional[TextAlign]:
```

File: source/NVDAObjects/window/excel.py [hunk 2]
```diff
@@ -54,28 +54,48 @@
...
+xlVAlignJustify = -4130
+xlVAlignTop = -4160
+
 xlDown=-4121
 xlToLeft=-4159
 xlToRight=-4161
 xlUp=-4162
 xlCellWidthUnitToPixels = 7.5919335705812574139976275207592
 xlSheetVisible=-1
-alignmentLabels={
```
Comment (seanbudd) [inline source/NVDAObjects/window/excel.py:70]: this is an API breaking change

File: source/NVDAObjects/window/excel.py [hunk 3]
```diff
@@ -54,28 +54,48 @@
...
+xlHAlignLeft = -4131
+xlHAlignRight = -4152
+
+# XlVAlign enumeration from https://docs.microsoft.com/en-us/office/vba/api/excel.xlvalign
+xlVAlignBottom = -4107
+xlVAlignCenter = -4108
+xlVAlignDistributed = -4117
+xlVAlignJustify = -4130
+xlVAlignTop = -4160
+
```
Comment (seanbudd) [inline source/NVDAObjects/window/excel.py:74]: can these become 2 separate `IntEnum`s? `horizontalAlignmentLabels` and `verticalAlignmentLabels` can be their displayStringLabels

File: source/controlTypes/formatFields.py [hunk 4]
```diff
@@ -99,3 +99,83 @@ def translateFromAttribute(fontSize: str) -> str:
...
+	# Translators: Reported when text is right-aligned.
+	TextAlign.RIGHT: _("align right"),
+	# Translators: Reported when text is justified.
+	# See http://en.wikipedia.org/wiki/Typographic_alignment#Justified
+	TextAlign.JUSTIFY: _("align justify"),
+	# Translators: Reported when text is justified with character spacing (Japanese etc)
+	# See http://kohei.us/2010/01/21/distributed-text-justification/
+	TextAlign.DISTRIBUTE: _("align distributed"),
+	# Translators: Reported when text is centered across multiple cells.
+	TextAlign.CENTER_ACROSS_SELECTION: _("align center across multiple cells"),
```
Comment (seanbudd) [inline source/controlTypes/formatFields.py:140]: Should this say "centered across selection" instead? it's shorter and handles the generic case better. we may want to use `CENTER_ACROSS_SELECTION` in other contexts to excel

**Comment by seanbudd** [comment]
Thanks @CyrilleB79 - this is almost ready

**Comment by CyrilleB79** [comment]
@seanbudd all the review comments have been implemented; this PR is ready again.

File: source/NVDAObjects/window/excel.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 import ctypes
 import enum
 from typing import (
-	Optional, Dict,
+	Optional, Dict, Any
```
Comment (seanbudd) [inline source/NVDAObjects/window/excel.py:11]:
```suggestion
	Any,
	Dict,
	Optional,
```

File: source/NVDAObjects/window/excel.py [hunk 6]
```diff
@@ -48,35 +48,103 @@
...
-xlJustify=-4130
-xlLeft=-4131
-xlRight=-4152
-xlDistributed=-4117
-xlBottom=-4107
-xlTop=-4160
+
+
+class XlHAlign(DisplayStringIntEnum):
+	# XlHAlign enumeration  from https://docs.microsoft.com/en-us/office/vba/api/excel.xlhalign
```
Comment (seanbudd) [inline source/NVDAObjects/window/excel.py:62]:
```suggestion
	# XlHAlign enumeration from https://docs.microsoft.com/en-us/office/vba/api/excel.xlhalign
```

**Review by seanbudd — APPROVED** [review]
Thanks, I'll get this merged soon

**Pull request merged**